pkg/sentry/devices/memdev/kmsg.go

```diff
@@ -0,0 +1,161 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memdev
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/safemem"
+	"gvisor.dev/gvisor/pkg/sentry/vfs"
+	"gvisor.dev/gvisor/pkg/syserror"
+	"gvisor.dev/gvisor/pkg/usermem"
+	"gvisor.dev/gvisor/pkg/waiter"
+)
+
+const (
+	// BufferEntryMax is the maximum number of entries retained by kmsg before new entries overwrite old ones.
+	BufferEntryMax = 512
+
+	// BufferMax is the maximum size of an individual entry, in bytes.
+	BufferMax = 1024
+)
+
+const kmsgDevMinor = 11
+
+// kmsgDevice implements vfs.Device for /dev/kmsg.
+type kmsgDevice struct{}
+
+// Open implements vfs.Device.Open.
+func (kmsgDevice) Open(ctx context.Context, mnt *vfs.Mount, vfsd *vfs.Dentry, opts vfs.OpenOptions) (*vfs.FileDescription, error) {
+	fd := &kmsgFD{}
+	if err := fd.vfsfd.Init(fd, opts.Flags, mnt, vfsd, &vfs.FileDescriptionOptions{
+		UseDentryMetadata: true,
+	}); err != nil {
+		return nil, err
+	}
+	return &fd.vfsfd, nil
+}
+
+// kmsgFD implements vfs.FileDescriptionImpl for /dev/kmsg.
+type kmsgFD struct {
+	vfsfd vfs.FileDescription
+	vfs.FileDescriptionDefaultImpl
+	vfs.DentryMetadataFileDescriptionImpl
+	vfs.NoLockFD
+
+	// buffer is a circular buffer that store user's write to kmsg.
+	m          sync.Mutex
+	buffer     [BufferEntryMax]*bytes.Buffer
+	readIndex  int
+	writeIndex int
+	// bufferLength is the number of unread entries in the buffer
+	bufferLength int
+}
+
+// Release implements vfs.FileDescriptionImpl.Release.
+func (fd *kmsgFD) Release() {
+}
+
+// Read implements vfs.FileDescriptionImpl.Read.
+func (fd *kmsgFD) Read(ctx context.Context, dst usermem.IOSequence, opts vfs.ReadOptions) (int64, error) {
+	fd.m.Lock()
+	defer fd.m.Unlock()
+
+	// When an entry gets overwritten in the circular buffer, next read()
+	// will return EPIPE and move readIndex to the next available record.
+	// See Documentation/ABI/testing/dev-kmsg in the Linux source for reference.
+	if fd.bufferLength > BufferEntryMax {
+		fd.readIndex = fd.writeIndex
+		fd.bufferLength = len(fd.buffer)
+		return 0, syserror.EPIPE
+	}
+	if fd.bufferLength == 0 {
+		if fd.vfsfd.StatusFlags()&^linux.O_NONBLOCK != 0 {
+			return 0, syserror.EAGAIN
+		}
+		return 0, syserror.ErrWouldBlock
+	}
+	if fd.buffer[fd.readIndex].Len() > int(dst.NumBytes()) {
+		return 0, syserror.EINVAL
+	}
+	bytesCopied, err := dst.CopyOutFrom(ctx, safemem.FromIOReader{bytes.NewReader(fd.buffer[fd.readIndex].Bytes())})
```

Comment:
Aside: using the built-in types will be a lot faster since they implementation all the safemem primitives directly.